File: C/Merge_Sort.c
#include <stdio.h>

// declare functions implemented after main
void printArr(int *arr, int n);
void mergeSort(int *arr, int left, int right);
void merge(int *arr, int left, int mid, int right);

int main()
{
	// initialize array and get length
	int arr[] = { 54, 45, 75, 1, 324, 3, 2, 4, 33, 2 };
	int n = sizeof(arr) / sizeof(int);

	// print initial array, sort array, then print sorted array
	printArr(arr, n);
	mergeSort(arr, 0, n - 1);
	printArr(arr, n);
	return 0;
}

// sort array using the merge sort algorithm
void mergeSort(int *arr, int left, int right)
{
	int mid;

	// sort left subarray and right subarray, then merge the two
	if (left < right)
	{
		mid = (left + right) / 2;
		mergeSort(arr, left, mid);
		mergeSort(arr, mid + 1, right);
		merge(arr, left, mid, right);
	}
}

// merge two sorted arrays
void merge(int *arr, int left, int mid, int right)
{
	// get length of each subarray
	int nLeft = mid - left + 1;
	int nRight = right - mid;
	
	// create each subarray
	int *leftArr = (int *)malloc(sizeof(int) * nLeft);
	int *rightArr = (int *)malloc(sizeof(int) * nRight);
	for (int i = 0; i < nLeft; i++)
		leftArr[i] = arr[left + i];
	for (int i = 0; i < nRight; i++)
		rightArr[i] = arr[mid + 1 + i];

	// initialize variables to track each subarray's index
	int iLeft = 0;
	int iRight = 0;
	int iArr = left;

	while (1)
	{
		// if either subarray is empty, place all remaining elements of
		// the other subarray into arr
		if (iLeft == nLeft)
		{
			for (int i = 0; i < nRight - iRight; i++)
			{
				arr[iArr + i] = rightArr[iRight + i];
			}
			break;
		}
		else if (iRight == nRight)
		{
			for (int i = 0; i < nLeft - iLeft; i++)
			{
				arr[iArr + i] = leftArr[iLeft + i];
			}
			break;
		}
		
		// compare first element of each subarray, place smaller element in arr
		if (leftArr[iLeft] < rightArr[iRight])
		{
			arr[iArr] = leftArr[iLeft];
			iLeft++;
		}
		else
		{
			arr[iArr] = rightArr[iRight];
			iRight++;
		}

		// increment arr index
		iArr++;
	}

	// free malloc'd space for two subarrays
	free(leftArr);
	free(rightArr);
}

// prints an array
void printArr(int *arr, int n)
{
	for (int i = 0; i < n; i++)
	{
		printf("%d ", arr[i]);
	}
	printf("\n");
}

File: C#/Merge_Sort.cs
using System;

namespace Sort
{
	class Program
	{
		static int Main()
		{
            // initialize array
            int[] arr = { 34, 1, 324, 6, 3, 12, 3, 254, 24};

            // print original array, sort, then print sorted array
            printArr(arr);
            mergeSort(arr, 0, arr.Length - 1);
            printArr(arr);
            return 0;
		}

        // sorts an array using a merge sort algorithm
        static void mergeSort(int[] arr, int low, int high)
        {
            if (low < high)
            {
                int mid = (high + low) / 2;

                // sort left side and right side, then merge them together
                mergeSort(arr, low, mid);
                mergeSort(arr, mid + 1, high);
                merge(arr, low, mid, high);
            }
        }

        // merges two sorted subarrays, which are defined by indicies "low",
        // "mid", and "high"
        static void merge(int[] arr, int low, int mid, int high)
        {
            // create new arrays for left and right subarrays
            int[] left = new int[mid - low + 1];
            int[] right = new int[high - mid];
            Array.Copy(arr, low, left, 0, mid - low + 1);
            Array.Copy(arr, mid + 1, right, 0, high - mid);

            // initialize main, left, and right array index counting variables
            int iArr = low;
            int iL = 0;
            int iR = 0;

            // compare first elements of each array, placing the smaller
            // element into the next position in the main array
            while (iL < left.Length && iR < right.Length)
            {
                if (left[iL] <= right[iR])
                {
                    arr[iArr] = left[iL];
                    iL++;
                }
                else
                {
                    arr[iArr] = right[iR];
                    iR++;
                }
                iArr++;
            }

            // place all elements from remaining array into main array
            if (iR == right.Length)
            {
                for (int i = iL; i < left.Length; i++)
                {
                    arr[iArr] = left[i];
                    iArr++;
                }
            }
            else
            {
                for (int i = iR; i < right.Length; i++)
                {
                    arr[iArr] = right[i];
                    iArr++;
                }
            }
        }

        // prints an array
        static void printArr(int[] arr)
        {
            int n = arr.Length;
            for (int i = 0; i < n; i++)
            {
                Console.Write("{0} ", arr[i]);
            }
            Console.Write("\n");
        }
	}
}